File: C++/Leetcode/Hashmap/happyNumber.cpp
#include <iostream>
#include <set>

class Solution {
public:
    bool isHappy(int n) {
        std::set<int> seen;
        while (n != 1 && seen.find(n) == seen.end()) {
            seen.insert(n);
            int val_sum = 0;
            while (n != 0) {
                int digit = n % 10; // Extract the last digit
                val_sum += digit * digit;
                n /= 10; // Remove the last digit
            }
            n = val_sum;
        }
        return n == 1; // If n == 1, it's a happy number
    }
};

int main() {
    Solution solution_instance;
    std::cout << solution_instance.isHappy(19) << std::endl;
    return 0;
}

File: Python/Leetcode/HashMap/happyNumber.py
class Solution:
    def isHappy(self, n: int) -> bool:
        seen = set()
        while n != 1:
            if n in seen:
                return False
            else:
                seen.add(n)
                n = sum(int(digit) ** 2 for digit in str(n))
        return True
    
if __name__ == '__main__':
    solution_instance = Solution()
    print(solution_instance.isHappy(19))